In [ ]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.preprocessing import StandardScaler
from sklearn.cluster import KMeans

In [16]:
df = pd.read_excel('online_retail.xlsx')
df.head()

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
0,489434,85048,15CM CHRISTMAS GLASS BALL 20 LIGHTS,12,2009-12-01 07:45:00,6.95,13085.0,United Kingdom
1,489434,79323P,PINK CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
2,489434,79323W,WHITE CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
3,489434,22041,"RECORD FRAME 7"" SINGLE SIZE",48,2009-12-01 07:45:00,2.10,13085.0,United Kingdom
4,489434,21232,STRAWBERRY CERAMIC TRINKET BOX,24,2009-12-01 07:45:00,1.25,13085.0,United Kingdom


In [17]:
print("Size:", df.shape)
print("\nType:")
print(df.dtypes)
print("\nLeft:")
print(df.isnull().sum())

Size: (525461, 8)

Type:
Invoice                object
StockCode              object
Description            object
Quantity                int64
InvoiceDate    datetime64[ns]
Price                 float64
Customer ID           float64
Country                object
dtype: object

Left:
Invoice             0
StockCode           0
Description      2928
Quantity            0
InvoiceDate         0
Price               0
Customer ID    107927
Country             0
dtype: int64


In [ ]:
df.shape



(525461, 8)

In [18]:
df['Total Price'] = df['Quantity'] * df['Price']
df.head() 

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country,Total Price
0,489434,85048,15CM CHRISTMAS GLASS BALL 20 LIGHTS,12,2009-12-01 07:45:00,6.95,13085.0,United Kingdom,83.4
1,489434,79323P,PINK CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom,81.0
2,489434,79323W,WHITE CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom,81.0
3,489434,22041,"RECORD FRAME 7"" SINGLE SIZE",48,2009-12-01 07:45:00,2.10,13085.0,United Kingdom,100.8
4,489434,21232,STRAWBERRY CERAMIC TRINKET BOX,24,2009-12-01 07:45:00,1.25,13085.0,United Kingdom,30.0


In [19]:
avg_purchase = df['Total Price'].mean() 
print(f"Average purchase value (per line item): {avg_purchase:.2f}")

Average purchase value (per line item): 18.15


In [21]:
purchase_freq = df.groupby('Customer ID')['Invoice'].nunique()
print(f"Average purchase frequency (orders per customer): {purchase_freq.mean():.2f}")

Average purchase frequency (orders per customer): 5.38


In [22]:
customer_value = df.groupby('Customer ID')['Total Price'].sum()
print(f"Average Customer Lifetime Value: {customer_value.mean():.2f}")

Average Customer Lifetime Value: 1904.68


## Descriptive Statistics

The average purchase value per line item is £18.15. Customers place 
an average of 5.38 orders each, and the average Customer Lifetime 
Value (total spend per customer) is £1904.68 — though this average 
is likely skewed upward by a small number of very high-spending 
customers, which the clustering analysis below will help identify.